src/HTTP.jl

```diff
@@ -1,83 +1,586 @@
+
+
+AWS Authenticaiton options
+
+ - `awsauthorization = false`, enable AWS4 Authentication.
+ - `aws_service = split(uri.host, ".")[1]`
+ - `aws_region = split(uri.host, ".")[2]`
+ - `aws_access_key_id = ENV["AWS_ACCESS_KEY_ID"]`
+ - `aws_secret_access_key = ENV["AWS_SECRET_ACCESS_KEY"]`
+ - `aws_session_token = get(ENV, "AWS_SESSION_TOKEN", "")`
+ - `body_sha256 = digest(MD_SHA256, body)`,
+ - `body_md5 = digest(MD_MD5, body)`,
+
+
+Cookie options
+
+ - `cookies = false`, enable cookies.
+ - `cookiejar::Dict{String, Set{Cookie}}=default_cookiejar`
+
+
+Cananoincalization options
+
+ - `canonicalizeheaders = false`, rewrite request and response headers in
+   Canonical-Camel-Dash-Format.
+
+
+## Request Body Examples
+
+String body:
+```julia
+request("POST", "http://httpbin.org/post", [], "post body data")
+```
+
+Stream body from file:
+```julia
+io = open("post_data.txt", "r")
+request("POST", "http://httpbin.org/post", [], io)
+```
+
+Generator body:
+```julia
+chunks = ("chunk\$i" for i in 1:1000)
+request("POST", "http://httpbin.org/post", [], chunks)
+```
+
+Collection body:
+```julia
+chunks = [preamble_chunk, data_chunk, checksum(data_chunk)]
+request("POST", "http://httpbin.org/post", [], chunks)
+```
+
+`open() do io` body:
+```julia
+HTTP.open("POST", "http://httpbin.org/post") do io
+    write(io, preamble_chunk)
+    write(io, data_chunk)
+    write(io, checksum(data_chunk))
 end
+```
 
-macro uninit(expr)
-    if !isdefined(Base, :uninitialized)
-        splice!(expr.args, 2)
-    end
-    return esc(expr)
+
+## Response Body Examples
+
+String body:
+```julia
+r = request("GET", "http://httpbin.org/get")
+println(String(r.body))
+```
+
+Stream body to file:
+```julia
+io = open("get_data.txt", "w")
+r = request("GET", "http://httpbin.org/get", response_stream=io)
+close(io)
+println(read("get_data.txt"))
+```
+
+Stream body through buffer:
+```julia
+io = BufferStream()
+@async while !eof(io)
+    bytes = readavailable(io))
+    println("GET data: \$bytes")
 end
+r = request("GET", "http://httpbin.org/get", response_stream=io)
+close(io)
+```
 
-if !isdefined(Base, :pairs)
-    pairs(x) = x
+Stream body through `open() do io`:
+```julia
+r = HTTP.open("GET", "http://httpbin.org/stream/10") do io
+   while !eof(io)
+       println(String(readavailable(io)))
+   end
 end
 
-if !isdefined(Base, :Nothing)
-    const Nothing = Void
-    const Cvoid = Void
+HTTP.open("GET", "https://tinyurl.com/bach-cello-suite-1-ogg") do http
+    n = 0
+    r = startread(http)
+    l = parse(Int, header(r, "Content-Length"))
+    open(`vlc -q --play-and-exit --intf dummy -`, "w") do vlc
+        while !eof(http)
+            bytes = readavailable(http)
+            write(vlc, bytes)
+            n += length(bytes)
+            println("streamed \$n-bytes \$((100*n)÷l)%\\u1b[1A")
+        end
+    end
 end
+```
+
+
+## Request and Response Body Examples
+
+String bodies:
+```julia
+r = request("POST", "http://httpbin.org/post", [], "post body data")
```

Comment:
Here as well, I'd really like to encourage users to always do `HTTP.request`.

Reply:
cb1b8505eef94916f451e302a8665b02c432c48d